### ISSUE TITLE
Incorrect memory alignment of components.

### ISSUE THREAD

**Issue by RobertBiehl** [SanderMertens/flecs#478]
**Describe the bug**
Component structs passed to system do not adhere to the defined memory alignment correctly.

**To Reproduce**

1. enable avx

2. Create component with aligned struct, e.g.
```
#include <cglm/cglm.h>
typedef struct EcsTransform {
    CGLM_ALIGN_MAT mat4 value;
} EcsTransform;
```

3. Inside system check address of pointer to struct

```
void TestSystem(ecs_iter_t* it) {
    EcsTransform3 *transform = ecs_term(it, EcsTransform3, 1);
    printf("(EcsTransform3 *) addr = %#010x", transform);

}
```
Sometimes prints addresses not aligned to 32 bytes as expected from CGLM_ALIGN_MAT.
This causes crashes in calls to AVX intrinsics.

**Expected behavior**
Address of component should be aligned correctly, in this case to 32 bytes.

**Additional context**
Add any other context about the problem here (operating system, hardware, ...).

**Comment by SanderMertens** [comment]
Hm, flecs should be able to (and has been verified to) handle alignment correctly. It does rely on `alignof/ECS_ALIGNOF` reporting the correct alignment, could you verify that if you do `ECS_ALIGNOF(EcsTransform3)` it returns 32?

Additionally you could run the following code to check if the correct alignment has been registered with flecs:
```c
ECS_COMPONENT(world, EcsTransform3);

EcsComponent *ptr = ecs_get(world, ecs_id(EcsTransform3), EcsComponent);
printf("%d\n", ptr->alignment);
```

**Comment by chengts95** [comment]
Yes, I have the same problem, I was thinking it was my fault, but the alignment of components is always 16bit even I changed the malloc function to use default 32bit alignment.

**Comment by SanderMertens** [comment]
@RobertBiehl @chengts95 if you're interested in debugging this further, the part of the code that should take care of the alignment is `ecs_vector_t`. 

A vector is a contiguous block of memory with a header in front of it (containing the number of elements + allocated size). At an offset from the header the actual array begins, and it's important that this offset is aligned to the alignment of the component type.

The alignment is computed by macro's and passed into the vector functions (to allow for compile-time optimization where the vector type is known). If anything's wrong with how the component alignment is used, I would expect that to be in this code.

I'd like to look into this myself as well but it may take a bit of time before I can get to it.

**Comment by chengts95** [comment]
If I am right, the alignment attribute can only determine the padding of the struct to fulfill the address alignment requirement, like expanding a 14byte struct to 16byte so its array is aligned.  
However, the heap memory address is determined by the malloc function, the malloc function is default to 16bit and I didn't see the vector is created by aligned_alloc anywhere, so it will be 16bit aligned.  So @RobertBiehl's struct cannot ensure a 32bit aligned address.

My purpose was to make components with double float-point numbers aligned to 32bit addresses so aligned AVX can be used, which means the first double number's address should be aligned to 32bit. The key problem is to have the starting address aligned to 32bit even the struct itself doesn't have an alignment attribute.

**Comment by chengts95** [comment]
> @RobertBiehl @chengts95 if you're interested in debugging this further, the part of the code that should take care of the alignment is `ecs_vector_t`.
> 
> A vector is a contiguous block of memory with a header in front of it (containing the number of elements + allocated size). At an offset from the header the actual array begins, and it's important that this offset is aligned to the alignment of the component type.
> 
> The alignment is computed by macro's and passed into the vector functions (to allow for compile-time optimization where the vector type is known). If anything's wrong with how the component alignment is used, I would expect that to be in this code.
> 
> I'd like to look into this myself as well but it may take a bit of time before I can get to it.

The size of the current header is 8 or 16byte. I don't think one header type can have different padding and where is the code to compute the offset?

**Comment by SanderMertens** [comment]
@chengts95 this is the code that computes the offset: https://github.com/SanderMertens/flecs/blob/master/include/flecs/private/vector.h#L50

Now that I'm looking at the code I think I also see the issue 💥 . The offset calculation simply does `max(header_size, alignment)` which works as long as the alignment is not larger than the alignment of the `malloc`, but in this case (32 bits alignment vs. 16 bits malloc alignment) this won't work.

If malloc returns address 16 then the offset calculation would add 32 which ends up being 48 which is unaligned. The code must be changed to take the max of `header_size` and the starting address _aligned to_ the component alignment.

**Comment by chengts95** [comment]
> @chengts95 this is the code that computes the offset: https://github.com/SanderMertens/flecs/blob/master/include/flecs/private/vector.h#L50
> 
> Now that I'm looking at the code I think I also see the issue 💥 . The offset calculation simply does `max(header_size, alignment)` which works as long as the alignment is not larger than the alignment of the `malloc`, but in this case (32 bits alignment vs. 16 bits malloc alignment) this won't work.
> 
> If malloc returns address 16 then the offset calculation would add 32 which ends up being 48 which is unaligned. The code must be changed to take the max of `header_size` and the starting address _aligned to_ the component alignment.

That is! I also want to have a 32byte aligned double array, if we determine the alignment only by the struct size, you cannot get an AVX array with a plain double number component, if we can manually choose the alignment just like malloc and aligned_alloc, it will be better! Normally people can manually peel the loop to fit the AVX requirement, but it is more convenient if the starting address is aligned.

**Comment by SanderMertens** [comment]
Sorry for the late reply! I've been looking at ways to address this, but there isn't a straightforward way to solve this with the current vector without changing lots of code.

I'll soon start working again on more storage related stuff, and one thing I've been wanting to do is change the component storage from a regular vector to a paged vector. I'll make sure that the new storage is able to deal with >16bits alignment.

**Comment by nimble0** [comment]
Is it possible to add an assert that a component will be aligned correctly?

**Comment by chengts95** [comment]
I feel the starting address alignment problem is not a very big deal in most cases.  Unaligned memory load/store operations are not very cirtical in common cases. For simple float-point data arrays, you can peel the unaligned part and perform avx2 operations only on the aligned memory. The fact is most ECS frameworks do not even allow users to access the chunks of archetypes in raw memory, which removes the benefits of ECS for computation. Flecs is the only one I have known to make this correct.

To solve this problem we have to use C11 or later. Otherwise there is no `alignas` ,`alignof` and `aligned_alloc`.

**Comment by addbounty** [comment]
[![banner](https://bountybot.dev/api/display/github/SanderMertens/flecs/478/banner?v=86694d6e-552f-40f2-8177-f01452133adf)](https://bountybot.dev/github/SanderMertens/flecs/issues/478)
[![button](https://bountybot.dev/api/display/github/SanderMertens/flecs/478/button?v=86694d6e-552f-40f2-8177-f01452133adf)](https://bountybot.dev/github/SanderMertens/flecs/issues/478)

Make a Draft PR **early** so others can see you are working on it! To automatically create one:

```sh
# Using npx (installed if you have NodeJS/npm)
npx bountybot solve SanderMertens/flecs#478
```

[Join Our Discord](https://discord.gg/jcQpX6jW8u) to connect with other bounty hunters and get help.

When merged, you will receive the bounty!

**Comment by DominikGrabiec** [comment]
Have there been any more progress on this issue?

At work we want to use AVX2 with Eigen types stored within flecs but run into alignment assertions from Eigen when constructing the elements in the component array as it isn't sufficiently aligned. Using C++ we see that the class is correctly aligned to 32 bytes, the alignment in the type info struct that gets sent to the ctor function is correctly set to 32 bytes, but the object ends up being emplaced with placement new into memory that is aligned/offset to 16 bytes which then triggers an assert within Eigen. From the looks of it the array pointer assigned to the column is not aligned to the requirements of the type, so all elements would not be aligned correctly.

I saw in the pull request [#1110](https://github.com/SanderMertens/flecs/pull/1110) that the idea is to use pluggable storage, but there doesn't seem to have been any development on it for a while.

Is the only work around to recompile flecs with `FLECS_USE_OS_ALLOC` and always feed it with memory aligned to the maximum required? Is there no way to do a per-type aligned memory allocation?

**Comment by chengts95** [comment]
> Have there been any more progress on this issue?
> 
> At work we want to use AVX2 with Eigen types stored within flecs but run into alignment assertions from Eigen when constructing the elements in the component array as it isn't sufficiently aligned. Using C++ we see that the class is correctly aligned to 32 bytes, the alignment in the type info struct that gets sent to the ctor function is correctly set to 32 bytes, but the object ends up being emplaced with placement new into memory that is aligned/offset to 16 bytes which then triggers an assert within Eigen. From the looks of it the array pointer assigned to the column is not aligned to the requirements of the type, so all elements would not be aligned correctly.
> 
> I saw in the pull request [#1110](https://github.com/SanderMertens/flecs/pull/1110) that the idea is to use pluggable storage, but there doesn't seem to have been any development on it for a while.
> 
> Is the only work around to recompile flecs with `FLECS_USE_OS_ALLOC` and always feed it with memory aligned to the maximum required? Is there no way to do a per-type aligned memory allocation?

If we want flecs to support aligned allocation, it is hard because aligned malloc is not included in c99 which is flecs's target. But it is possible to override os malloc api function on your own.

**Comment by SanderMertens** [comment]
I'll take a look at whether this is something I can fix for the next release.
